[core] Add specific tensor descriptor for Result

praasz opened the pull request:
### Details:
 - Introduce shared tensor descriptor:
   - It has specific tensor names.
   - When specific tensor names set they will be used as tensor names otherwise the shred tensor names are used.
   - Specific tensor names will be added to shared descriptor (input descriptor), and transferred to new tensor descriptor if changed.
   
 - Use shared tensor descriptor for Result, which got specific tensor names.
 - Add dev API Extension interface for descriptor::Tensor for internal OV use.
 - Update `ReverseShapeAndTypeInfer` to use descriptor interface instead change properties directly. Update `PartialShape` of descriptor can change it in undefined state with `Shape` property.
 - Improves tensor names handling for model outputs introduced in #23713
 - Fix tensor names issue in `unroll_if` transformation (ticket: 114763)

### Tickets:
 - CVS-143105
 - CVS-146649
 - CVS-114763

### Blocked  by:
- intel-innersource/frameworks.ai.openvino.tests#1684 - fixes e2e tests

praasz committed:
Use pimpl in Tensor descriptor to hide implementation details from public API

praasz committed:
Use pimpl in Tensor descriptor to hide implementation details from public API

praasz committed:
Use descriptor iface in ReverseShapeAndTypeInfer to update shapes

praasz committed:
Use descriptor iface in ReverseShapeAndTypeInfer to update shapes

praasz committed:
Don't use deprecated ctor for descriptor

praasz committed:
Don't use deprecated ctor for descriptor

praasz committed:
Export `TensorExtension::Hasher` and `TensorExtension::Equal`

praasz committed:
Export `TensorExtension::Hasher` and `TensorExtension::Equal`

praasz committed:
Update preprocess build of OutputInfo

praasz committed:
Update preprocess build of OutputInfo

praasz committed:
Fix preserve rt_info for Result

praasz committed:
Fix preserve rt_info for Result

praasz committed:
Fix mo API tensor names test

praasz committed:
Fix mo API tensor names test

praasz committed:
Update tensor descriptor compare in stateful to stateless  tests

praasz committed:
Update tensor descriptor compare in stateful to stateless  tests

praasz committed:
Update PPP output info dump

praasz committed:
Update PPP output info dump

praasz committed:
Update check output name in mo_convert tests

praasz committed:
Update check output name in mo_convert tests

praasz committed:
Correct add tensor names to Result

praasz committed:
Correct add tensor names to Result

praasz committed:
Test use only Result names if exists

praasz committed:
Test use only Result names if exists

praasz committed:
Update set result tensor names for v10

praasz committed:
Update set result tensor names for v10

praasz committed:
Result tensor uses shared tensor names

praasz committed:
Result tensor uses shared tensor names

rkazants commented on tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py:
```diff
@@ -940,14 +940,16 @@ def __call__(self, input):
+        # set model's output names will not change connected node names
+        # output can have additional names from connected input node
         ov_model.outputs[0].get_tensor().set_names({"name1"})
-        assert ov_model.outputs[0].names == {"name1"}
+        assert "name1" in ov_model.outputs[0].names
 
         ov_model.validate_nodes_and_infer_types()
-        assert ov_model.outputs[0].names == {"name1"}
+        assert "name1" in ov_model.outputs[0].names
 
         ov_model.outputs[0].get_tensor().set_names({"name2"})
-        assert ov_model.outputs[0].names == {"name2"}
+        assert "name2" in ov_model.outputs[0].names
```
why now tensor names is changed? We should have the single tensor name instead of several

rkazants reviewed (dismissed):
why several tensor names are not clear

praasz committed:
Merge branch 'master' into feature/specific-tensor-descriptor-for-results

praasz committed:
Merge branch 'feature/specific-tensor-descriptor-for-results' of github.com:praasz/openvino into feature/specific-tensor-descriptor-for-results

praasz committed:
Move shared tensro descriptor to separate file

praasz committed:
Add missing includes

praasz committed:
Update compare result tensor for stateful to stateless transformation test

praasz committed:
Merge branch 'master' into feature/specific-tensor-descriptor-for-results

praasz committed:
When Result tensor set will hide names from Result's input

praasz committed:
Add doxy comments

praasz committed:
Apply review comments

praasz commented on src/core/src/descriptor/tensor.cpp:
```diff
@@ -8,150 +8,294 @@
+    RTMap m_rt_map;
+    std::string m_legacy_name;
+
+    static decltype(m_name_it) find_new_any_name(const decltype(m_names)& names, decltype(m_name_it) name) {
```
Removed.

praasz commented on src/core/include/openvino/core/descriptor/tensor.hpp:
```diff
@@ -22,103 +22,107 @@ namespace ov {
            const PartialShape& pshape,
            const std::unordered_set<std::string>& names = {});
+
+    OPENVINO_DEPRECATED("This constructor is deprecated. Will be removed in 2025.0")
```
Changed to 2026.0

praasz commented on src/core/tests/type_prop/result.cpp:
```diff
@@ -51,3 +55,101 @@ TEST(type_prop, result_layout_invalid) {
+    EXPECT_THAT(result->output(0).get_names(), UnorderedElementsAre("input"));
+    EXPECT_THAT(result->get_output_tensor(0).get_names(), UnorderedElementsAre("input"));
+
+    EXPECT_THAT(result->output(0).get_names(), UnorderedElementsAre("input"));
```
No, removed.

slyalin commented on src/core/include/openvino/op/result.hpp:
```diff
@@ -13,6 +13,13 @@ namespace v0 {
+/// The Result operator output is special output which share tensor with node connected to this node.
+///
+/// The Result's output names are visible as model outputs names.
+/// When set/add Result's output tensor names they will treat as this result specific names which will be appended to
+/// connect input tensor or transferred to new tensor connect Result to new input.
+/// When Result has not specific names the names from connected input will be used.
```
It is hard to follow. Could you add more details and itemize all the variants?

praasz committed:
Merge branch 'master' into feature/specific-tensor-descriptor-for-results

jane-intel commented on tests/model_hub_tests/transformation_tests/test_stateful_to_stateless_transformation.py:
```diff
@@ -8,21 +8,46 @@
-        assert t1.get_element_type() == t2.get_element_type()
+def check_desc_tensors(expected_tensors, tensors):
+    # checks if tensors descriptors are same as expected
+    pattern_input = re.compile(R"input_restored.((past_key_values|present)\.(\d+)\.(key|value))")
```
Could you elaborate on why this change is needed? I'm a bit confused, are we backward compatible?

praasz replied:
Same as [here](https://github.com/openvinotoolkit/openvino/pull/25954#discussion_r1800909046)

jane-intel commented on tests/model_hub_tests/transformation_tests/test_stateful_to_stateless_transformation.py:
```diff
@@ -8,21 +8,46 @@
+    assert len(expected_tensors) == len(tensors)
+    for expected in expected_tensors:
+        # The `patch_stateful` in optimum use any name instead found key/value names OV will use names restore path
+        # Restore expected names to find tensor for compare (can be removed when HG optimum updated)
```
What is the problem with HG optimum? Does this mean that get_any_name returns different names now?

praasz replied:
The `get_any_name` is not reliable to use for testing to compare names when tensor can have many names or when some names are added to tensor.

As Result now not set its input names (they are not lost) the additional names can appear on compiled/transformed model. If there is many names get any name can give different name than on origin model and test fail but origin name is still on tensor.

This modification will only take stateless names on inputs and outputs and check if they are still present on model after transformation.

The HG issue was that it use any name and as there could be more names on tensor it pick the stateless pattern name or other. In case other name has been used the restore name path in transformation is used. This change will give more robust check whatever path is selected in transformation it will extract the names and prepare expected  and check if exists on input/outputs  on model after transformation.

The HG issue with any name has been fixed but OV can still use older version.

jane-intel commented on src/core/tests/type_prop/result.cpp:
```diff
@@ -51,3 +55,100 @@ TEST(type_prop, result_layout_invalid) {
+    const auto result = make_op(a);
+    result->output(0).set_names({"out"});
+
+    EXPECT_THAT(result->output(0).get_names(), UnorderedElementsAre("out"));
```
What is expected of result->input(0) here? Could you explain / add checks here?

praasz replied:
```cpp
EXPECT_THAT(result->input(0).get_tensor().get_names(), UnorderedElementsAre("out", "input a"));
```
Check will be added

jane-intel commented on src/core/src/preprocess/preprocess_impls.cpp:
```diff
@@ -325,11 +326,9 @@ void InputInfo::InputInfoImpl::dump(std::ostream& str,
-    const auto start_out_node_names = node.get_tensor().get_names();
-    node.get_tensor().set_names({});
-    result = std::dynamic_pointer_cast<opset8::Result>(node.get_node_shared_ptr());
+    const auto result = std::dynamic_pointer_cast<opset8::Result>(node.get_node_shared_ptr());
```
nit ov::as_type_ptr?

jane-intel commented on src/core/src/node.cpp:
```diff
@@ -472,8 +467,8 @@ ov::descriptor::Tensor& ov::Node::get_output_tensor(size_t i) const {
-    descriptor::Input input = m_inputs[i];
-    return input.get_tensor();
+    auto& input = m_inputs[i];
+    return input.get_output().get_tensor();
```
Why this had to be changed? `input.get_tensor()` wouldn't work?

praasz replied:
The first line it was just to remove unnecessary copy which was used to make a little hack (similar to `const_cast`) to remove consents from `input` as this function is `const`.
Without copy the connected node output tensor has to be returned which is not constant and it give better description what is returned . 

This is maybe separate topic but maybe `get_input_tensor` tensor should return const descriptor to avoid modification of descriptors which not belongs to this Node. The Node's inputs are not tensors owners.

e-nugmanova reviewed (commented):
Hello, thank you for this PR.
Could you share why CVS-115911 results weren't taken into account? It seems like the presentation and decision was made to have different approach.

Also, please clarify what happens with get_any_name -- why it starts to be incompatible?

praasz commented on src/core/src/preprocess/preprocess_impls.cpp:
```diff
@@ -369,56 +368,36 @@ void OutputInfo::OutputInfoImpl::build(ov::ResultVector& results) {
+        // Reset friendly name of input node to avoid names collision
+        // when there is at a new node inserted by post-processing steps
+        // If no new nodes are inserted by post-processing, then we need to preserve friendly name of input
+        // as it's required for old API correct work
```
For PPP compatibility mode it has to be kept (I don't want to change this functionality in this PR).

For IRv11 it can be considered as  a bug which is solved in [PR #23713](https://github.com/openvinotoolkit/openvino/pull/23713/files#diff-7c4d81a6d4bb639ff9af23a3a8f09cda9ec474f9fa1d6acf988c21b9c5ede74dR373). 

The Node's (layer) name should not be moved which was reported in [CVS-127482](https://jira.devtools.intel.com/browse/CVS-127482)

praasz committed:
Review suggestions

praasz commented:
> Hello, thank you for this PR. Could you share why [CVS-115911](https://jira.devtools.intel.com/browse/CVS-115911) results weren't taken into account? It seems like the presentation and decision was made to have different approach.
> 
> Also, please clarify what happens with get_any_name -- why it starts to be incompatible?
I've seen the CVS-115911 after I've done this PR, so I was not aware about it.

But I've done similar tries the best options was to apply changes in descriptor or make additional operator.
With Identity operator I've got some doubts (decide to try improve descriptor):

- It is required to add new operator
- There is add ne transformations to add/remove automatically between Result and other node
- When Identity added there has to be decision about current names which are Result's and which Result's input
- When Identity is removed the information which name was set to which Node is lost
- When Identity is connected to Result it could be required some dedicated handling from other nodes as Result is special

The current implementation is not making full names separation as Result's input has all names which allow to be backward compatible for (de)serialization.
To make full separation in descriptor will require addition effort and can be problematic for serialization.

Have this feature in descriptor then is not required to use additional operator (Identity can be useful for other purposes)
and there is no need to addition pre/post process model to add/remove them.

The functionality of `get_any_name` is not changed. It is not good to use in way that for some output on origin model we get any name then after model processing get again any name from same output then it maybe but must be same.
It depends how model is transformed as some outputs can get more names. Is better to `get_any_name` and check if this name is still in output after model processing.

praasz committed:
Merge remote-tracking branch 'origin/master' into feature/specific-tensor-descriptor-for-results

praasz committed:
Merge remote-tracking branch 'origin/master' into feature/specific-tensor-descriptor-for-results

praasz committed:
Enable UnrollIf tests on CPU
- add UnrollIf to check tensor names are not lost if body model is Parameter -> Result

Signed-off-by: Raasz, Pawel <pawel.raasz@intel.com>

praasz committed:
Merge branch 'master' into feature/specific-tensor-descriptor-for-results

alexsu52 reviewed (approved):
I confirm that this PR fixes NNCF layer-wise algorithms such as GPTQ.

praasz committed:
Merge remote-tracking branch 'origin/master' into feature/specific-tensor-descriptor-for-results

praasz committed:
Merge remote-tracking branch 'origin/master' into feature/specific-tensor-descriptor-for-results

praasz committed:
Merge branch 'master' into feature/specific-tensor-descriptor-for-results

praasz committed:
Merge branch 'master' into feature/specific-tensor-descriptor-for-results
